crates/ra_assists/src/handlers/number_representation.rs

```diff
@@ -0,0 +1,603 @@
+        }
+
+        f.write_str(&self.text)?;
+
+        if let Some(suffix) = &self.suffix {
+            f.write_str(suffix)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn identify_number_literal(literal: &ast::Literal) -> Option<NumberLiteral> {
+    match literal.kind() {
+        LiteralKind::IntNumber { suffix } => {
+            let token = literal.token();
+            let full_text = token.text().as_str();
+            let suffix_clone = suffix.clone();
+            let suffix_len = suffix.map(|s| s.len()).unwrap_or_default();
+            let non_suffix = &full_text[0..full_text.len() - suffix_len];
+            let maybe_prefix = if non_suffix.len() < 2 { None } else { Some(&non_suffix[0..2]) };
+            let (prefix, number_type) = match maybe_prefix {
+                Some("0x") => (maybe_prefix, NumberLiteralType::Hexadecimal),
+                Some("0b") => (maybe_prefix, NumberLiteralType::Binary),
+                Some("0o") => (maybe_prefix, NumberLiteralType::Octal),
+                _ => (None, NumberLiteralType::Decimal),
+            };
+            let prefix_len = prefix.map(|s| s.len()).unwrap_or_default();
+            let text = &non_suffix[prefix_len..];
+
+            let result = NumberLiteral {
+                number_type,
+                suffix: suffix_clone,
+                prefix: prefix.map(SmolStr::new),
+                text: text.to_string(),
+            };
+            Some(result)
+        }
+        _ => None,
+    }
+}
+
+fn is_int_number(literal: &ast::Literal) -> bool {
+    match literal.kind() {
+        LiteralKind::IntNumber { .. } => true,
+        _ => false,
+    }
+}
+
+fn remove_separator_from_string(str: &str) -> String {
+    str.replace("_", "")
+}
+
+pub(crate) fn remove_digit_separators(ctx: AssistCtx) -> Option<Assist> {
+    let literal = ctx.find_covering_node_at_offset::<ast::Literal>()?;
+    if !is_int_number(&literal) {
+        return None;
+    }
+
+    if !literal.syntax().text().contains_char('_') {
+        return None;
+    }
+
+    ctx.add_assist(AssistId("remove_digit_separators"), "Remove digit separators", |edit| {
+        edit.target(literal.syntax().text_range());
+        let new_text = remove_separator_from_string(&literal.syntax().text().to_string());
+        edit.replace(literal.syntax().text_range(), new_text);
+    })
+}
+
+fn len_without_separators(text: &str) -> usize {
+    let mut len = 0;
+    for c in text.chars() {
+        if c != '_' {
+            len += 1;
+        }
+    }
+    return len;
+}
+
+fn separate_number(text: &str, every: usize, digits_len: usize) -> String {
+    let mut result = String::with_capacity(digits_len + digits_len / every);
+    let offset = every - (digits_len % every);
+    let mut i = 0;
+    for c in text.chars() {
+        if c != '_' {
+            if (i != 0) && ((i + offset) % every == 0) {
+                result.push('_');
+            }
+            result.push(c);
+            i += 1;
+        }
+    }
```

Comment:
```suggestion
    for (i, c) in text.chars().enumerate() {
        if c != '_' {
            if (i != 0) && ((i + offset) % every == 0) {
                result.push('_');
            }
            result.push(c);
        }
    }
```

Reply:
This can't work because i should only increment for digits, not for separators (`'_'`) already in the string